backend/src/auth/tfa/tfa.controller.ts

```diff
@@ -1,90 +1,113 @@
 import {
+  BadRequestException,
   Body,
   Controller,
+  Get,
   Post,
-  Req,
-  Res,
+  Session,
   UnauthorizedException,
   UseGuards,
 } from '@nestjs/common';
-import { Request, Response } from 'express';
-import { User } from '../../core/entities';
-import { AuthenticatedGuard } from '../guards';
-import { TfaService } from './tfa.service';
+import {
+  AuthenticatedGuard,
+  TwoFactorDisabledGuard,
+  TwoFactorEnabledGuard,
+} from '../guards';
+import { TfaGenerateResponse, TfaService } from './tfa.service';
 import { TfaDto } from '../dto';
+import { User } from '../../core/entities';
+import { GetUser } from '../../core/decorators';
 
 @Controller('auth/tfa')
 @UseGuards(AuthenticatedGuard)
 export class TfaController {
   constructor(private readonly tfaService: TfaService) {}
 
-  @Post('generate')
-  async tfaGenerateSecret(@Res() res: Response, @Req() req: Request) {
-    const auth = await this.tfaService.tfaGenerateSecret(req.user as User);
-    res.setHeader('Content-Type', 'image/png');
-    console.log({ secret: auth.secret });
-    await this.tfaService.tfaGenerateQrCode(res, auth.otpAuthUrl);
+  @Get('generate')
+  @UseGuards(TwoFactorDisabledGuard)
+  async tfaGenerateSecret(@GetUser() user: User): Promise<TfaGenerateResponse> {
+    return await this.tfaService.tfaGenerateSecret(user);
   }
 
   @Post('enable')
-  async tfaEnable(@Req() req: Request, @Body() dto: TfaDto) {
-    if (!this.tfaService.tfaIsCodeValid(req.user as User, dto.tfaCode)) {
+  @UseGuards(TwoFactorDisabledGuard)
+  async tfaEnable(
+    @GetUser() user: User,
+    @Session() session: any,
```

Comment:
`any` -> Cries in typescript 😭

Reply:
HDFAUSFDHASUDH é que a sessão do objeto da request nao tem tipo definido mesmo, era melhor que do jeito que tava antes (`(session as any).field // ...`)